ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
+       * If the entry gets evicted later, we are still safe as we have saved all the metadata.
+       */
+      blobType = BlobType.MetadataBlob;
+      compositeBlobInfo = blobMetadata.getCompositeBlobInfo();
+      totalSize = compositeBlobInfo.getTotalSize();
+      chunkMetadataList = compositeBlobInfo.getChunkMetadataList();
+      blobInfo = blobMetadata.getBlobInfo();
+      lifeVersion = blobInfo.getLifeVersion();
+      serverBlobProperties = blobInfo.getBlobProperties();
+      chunkIndexToBuf = new ConcurrentHashMap<>();
+      chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+      successfullyDeserialized = true;
+      logger.info(String.format("Metadata cache-hit for blobId = %s", blobId));
+    }
+
+    /**
+     * Return the {@link MessageFormatFlags} to associate with the first getBlob chunk operation.
+     * @return {@link MessageFormatFlags#Blob} for {@link GetBlobOptions.OperationType#Data}, or
+     *         {@link MessageFormatFlags#All} by default.
+     */
+    @Override
+    MessageFormatFlags getOperationFlag() {
+      /**
+       * If we had a metadata cache hit, then we just need to get blobInfo to validate the cache entry.
+       * We don't have to fetch the entire metadata again.
+       */
+      return MessageFormatFlags.BlobInfo;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br>
+     * It would help to keep in mind while going through this method that the first chunk is either a metadata chunk
+     * or the only chunk of the blob.
+     */
+    @Override
+    void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInfo messageInfo)
+        throws IOException, MessageFormatException {
+
+      /**
+       * If we find blob metadata in frontend cache, validate it and use it.
+       */
+      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      if (!serverBlobProperties.toString().equals(receivedBlobProperties.toString())) {
+        /**
+         * This is a rare error case.
+         * If we find blob properties is in our frontend cache, we request just blobInfo from backend to validate
+         * the cache entry. If cached blob properties does not match received blob properties,
+         * we must delete the cache entry and throw a retriable exception to the user.
+         * The next request will re-fill the cache with a valid entry if the blob stills exists.
+         * We do not need to decrypt anything at this point.
+         */
+        String reason = "Cached blob property does not match received blob property";
+        deleteMetadata(reason);
+        logger.error(
+            String.format("%s for blobId = %s, cached blob property = %s, received blob property = %s", reason, blobId,
+                serverBlobProperties.toString(), receivedBlobProperties));
+        throw new MessageFormatException(reason, MessageFormatErrorCodes.IO_Error);
```

Comment:
we shouldn't throw MessageFormatException here.

Let's figure out why there is a difference between cached blob properties and requested blob properties. The only case i can think of is ttl.

Rather than that, I can't think so any other case. 

If we updated ttl  and they are still different, then something is probably very wrong.

We should terminate this operation by setting chunkException. And MessageFormatException is not a good one, we should use RouterException.

```
      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
      updateTtlIfRequired(serverBlobProperties, messageInfo);
      lifeVersion = messageInfo.getLifeVersion;
      if (!serverBlobProperties.equals(receivedBlobProperties)) {
            deleteMetadataInCache(reason);
            setChunkException(new RouterException(reason, RouterErrorCode.UnexpectedInternalError));
            return;
      }
```

Reply:
And set `chunkCompleted` to true as well.

Reply:
should we update lifeVersion here using `blobInfo.setLifeVersion(lifeVersion);`